cdap-app-fabric/src/main/java/io/cdap/cdap/internal/capability/CapabilityApplier.java

```diff
@@ -0,0 +1,402 @@
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import javax.annotation.Nullable;
+
+/**
+ * Class that applies capabilities
+ */
+public class CapabilityApplier {
+
+  private static final Logger LOG = LoggerFactory.getLogger(CapabilityApplier.class);
+  private static final Gson GSON = new Gson();
+  private static final int RETRY_LIMIT = 5;
+  private static final int RETRY_DELAY = 5;
+  private static final String CAPABILITY = "capability:%s";
+  private static final String APPLICATION = "application";
+  private static final ProgramTerminator NOOP_PROGRAM_TERMINATOR = programId -> {
+    // no-op
+  };
+  private final SystemProgramManagementService systemProgramManagementService;
+  private final ApplicationLifecycleService applicationLifecycleService;
+  private final ProgramLifecycleService programLifecycleService;
+  private final NamespaceAdmin namespaceAdmin;
+  private final CapabilityReader capabilityReader;
+  private final CapabilityWriter capabilityWriter;
+  private final MetadataSearchClient metadataSearchClient;
+
+  @Inject
+  CapabilityApplier(CConfiguration cConf, SystemProgramManagementService systemProgramManagementService,
+                    ApplicationLifecycleService applicationLifecycleService, NamespaceAdmin namespaceAdmin,
+                    ProgramLifecycleService programLifecycleService, CapabilityReader capabilityReader,
+                    CapabilityWriter capabilityWriter, DiscoveryServiceClient discoveryClient) {
+    this.systemProgramManagementService = systemProgramManagementService;
+    this.applicationLifecycleService = applicationLifecycleService;
+    this.programLifecycleService = programLifecycleService;
+    this.capabilityReader = capabilityReader;
+    this.capabilityWriter = capabilityWriter;
+    this.namespaceAdmin = namespaceAdmin;
+    this.metadataSearchClient = new MetadataSearchClient(discoveryClient);
+  }
+
+  /**
+   * Applies the given capability configurations
+   *
+   * @param capabilityConfigs
+   */
+  public void apply(Collection<? extends CapabilityConfig> capabilityConfigs) throws Exception {
+    List<CapabilityConfig> newConfigs = new ArrayList<>(capabilityConfigs);
+    Set<CapabilityConfig> enableSet = new HashSet<>();
+    Set<CapabilityConfig> disableSet = new HashSet<>();
+    Set<CapabilityConfig> deleteSet = new HashSet<>();
+    Map<String, CapabilityStatusRecord> currentCapabilities = capabilityReader.getAllCapabilities().stream().collect(
+      Collectors.toMap(CapabilityStatusRecord::getCapability, Function.identity()));
+    Map<String, CapabilityOperationRecord> currentOperations = capabilityReader.getCapabilityOperations().stream()
```

Comment:
The `getCapabilityOperations` should be only in the store class/interface. Generally think of the store as the internal class, while reader and writer are public. Since operation is internal to capability implementation, there is no need to expose it to either reader/writer interface

Reply:
Done.